swim/disseminator.go

```diff
@@ -109,63 +109,95 @@ func (d *disseminator) FullSync() (changes []Change) {
 	return changes
 }
 
-func (d *disseminator) IssueAsSender() []Change {
-	return d.issueChanges(nil)
+// IssueAsSender collects all changes a node needs when sending a ping or
+// ping-req. The second return value is a callback that raises the piggyback
+// counters of the given changes.
+func (d *disseminator) IssueAsSender() (changes []Change, bumpPiggybackCounters func()) {
+	changes = d.issueChanges()
+	return changes, func() {
+		d.bumpPiggybackCounters(changes)
+	}
 }
 
-func (d *disseminator) IssueAsReceiver(senderAddress string,
-	senderIncarnation int64, senderChecksum uint32) ([]Change, bool) {
+func (d *disseminator) bumpPiggybackCounters(changes []Change) {
+	d.Lock()
+	for _, change := range changes {
+		c, ok := d.changes[change.Address]
+		if !ok {
+			continue
+		}
 
-	filter := func(c *pChange) bool {
-		return senderIncarnation == c.SourceIncarnation &&
-			senderAddress == c.Source
+		c.p += 1
+		if c.p >= d.maxP {
+			delete(d.changes, c.Address)
+		}
 	}
+	d.Unlock()
+}
 
-	changes := d.issueChanges(filter)
+// IssueAsReceiver collects all changes a node needs when responding to a ping
+// or ping-req. Unlike IssueAsSender, IssueAsReceiver automatically increments
+// the piggyback counters. The second return value indicates whether a full sync
```

Comment:
Good work adding comments. This has helped me a lot.

I'd be even more liberal with this. E.g.:

> Unlike IssueAsSender, IssueAsReceiver automatically increments
> the piggyback counters.

Great highlight of an important detail. I'd take it one step further and explain _why_ it does this.

Anyway, this is not a blocker for merging.